[BUG] org.opensearch.search.pit.DeletePitMultiNodeIT#testDeleteWhileSearch

**Describe the bug**
In this test case, we set `deleted` flag after completing the execution of `DeletePitAction`(https://github.com/opensearch-project/OpenSearch/blob/ebdffbb2bf057096028aa3e6321be8a2ebc08ec4/server/src/internalClusterTest/java/org/opensearch/search/pit/DeletePitMultiNodeIT.java#L288), and based on the flag, we check whether the exception returned by the search request is allowed. But we cannot guarantee the `deleted` flag will be set immediately after specified pit is deleted.

```
WARNING: Uncaught exception in thread: Thread[opensearch[node_s_0][search],5,TGRP-DeletePitMultiNodeIT]
java.lang.AssertionError: java.util.concurrent.ExecutionException: Failed to execute phase [dfs], all shards failed; shardFailures {[SPCbwF64S3aUTy5JkzJBeg][index][0]: RemoteTransportException[[node_s0][127.0.0.1:57976][indices:data/read/search[phase/dfs]]]; nested: SearchContextMissingException[No search context found for id [1]]; }{[FyWJ9y7aRp2rNxpo6zbiRA][index][1]: RemoteTransportException[[node_s1][127.0.0.1:57974][indices:data/read/search[phase/dfs]]]; nested: SearchContextMissingException[No search context found for id [1]]; }{[SPCbwF64S3aUTy5JkzJBeg][index][2]: RemoteTransportException[[node_s0][127.0.0.1:57976][indices:data/read/search[phase/dfs]]]; nested: SearchContextMissingException[No search context found for id [2]]; }{[SPCbwF64S3aUTy5JkzJBeg][index][3]: RemoteTransportException[[node_s0][127.0.0.1:57976][indices:data/read/search[phase/dfs]]]; nested: SearchContextMissingException[No search context found for id [3]]; }{[SPCbwF64S3aUTy5JkzJBeg][index][4]: RemoteTransportException[[node_s0][127.0.0.1:57976][indices:data/read/search[phase/dfs]]]; nested: SearchContextMissingException[No search context found for id [4]]; }
	at __randomizedtesting.SeedInfo.seed([E454A72DB6FD2200]:0)
	at org.opensearch.search.pit.DeletePitMultiNodeIT.lambda$testDeleteWhileSearch$1(DeletePitMultiNodeIT.java:280)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.util.concurrent.ExecutionException: Failed to execute phase [dfs], all shards failed; shardFailures {[SPCbwF64S3aUTy5JkzJBeg][index][0]: RemoteTransportException[[node_s0][127.0.0.1:57976][indices:data/read/search[phase/dfs]]]; nested: SearchContextMissingException[No search context found for id [1]]; }{[FyWJ9y7aRp2rNxpo6zbiRA][index][1]: RemoteTransportException[[node_s1][127.0.0.1:57974][indices:data/read/search[phase/dfs]]]; nested: SearchContextMissingException[No search context found for id [1]]; }{[SPCbwF64S3aUTy5JkzJBeg][index][2]: RemoteTransportException[[node_s0][127.0.0.1:57976][indices:data/read/search[phase/dfs]]]; nested: SearchContextMissingException[No search context found for id [2]]; }{[SPCbwF64S3aUTy5JkzJBeg][index][3]: RemoteTransportException[[node_s0][127.0.0.1:57976][indices:data/read/search[phase/dfs]]]; nested: SearchContextMissingException[No search context found for id [3]]; }{[SPCbwF64S3aUTy5JkzJBeg][index][4]: RemoteTransportException[[node_s0][127.0.0.1:57976][indices:data/read/search[phase/dfs]]]; nested: SearchContextMissingException[No search context found for id [4]]; }
	at org.opensearch.common.util.concurrent.BaseFuture$Sync.getValue(BaseFuture.java:286)
	at org.opensearch.common.util.concurrent.BaseFuture$Sync.get(BaseFuture.java:273)
	at org.opensearch.common.util.concurrent.BaseFuture.get(BaseFuture.java:104)
	at org.opensearch.search.pit.DeletePitMultiNodeIT.lambda$testDeleteWhileSearch$1(DeletePitMultiNodeIT.java:270)
	... 1 more
Caused by: Failed to execute phase [dfs], all shards failed; shardFailures {[SPCbwF64S3aUTy5JkzJBeg][index][0]: RemoteTransportException[[node_s0][127.0.0.1:57976][indices:data/read/search[phase/dfs]]]; nested: SearchContextMissingException[No search context found for id [1]]; }{[FyWJ9y7aRp2rNxpo6zbiRA][index][1]: RemoteTransportException[[node_s1][127.0.0.1:57974][indices:data/read/search[phase/dfs]]]; nested: SearchContextMissingException[No search context found for id [1]]; }{[SPCbwF64S3aUTy5JkzJBeg][index][2]: RemoteTransportException[[node_s0][127.0.0.1:57976][indices:data/read/search[phase/dfs]]]; nested: SearchContextMissingException[No search context found for id [2]]; }{[SPCbwF64S3aUTy5JkzJBeg][index][3]: RemoteTransportException[[node_s0][127.0.0.1:57976][indices:data/read/search[phase/dfs]]]; nested: SearchContextMissingException[No search context found for id [3]]; }{[SPCbwF64S3aUTy5JkzJBeg][index][4]: RemoteTransportException[[node_s0][127.0.0.1:57976][indices:data/read/search[phase/dfs]]]; nested: SearchContextMissingException[No search context found for id [4]]; }
	at org.opensearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:677)
	at org.opensearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:373)
	at org.opensearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:716)
	at org.opensearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:485)
	at org.opensearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:295)
	at org.opensearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:75)
	at org.opensearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:749)
	at org.opensearch.transport.TransportService$6.handleException(TransportService.java:875)
	at org.opensearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1483)
	at org.opensearch.transport.InboundHandler.lambda$handleException$3(InboundHandler.java:421)
	at org.opensearch.common.util.concurrent.OpenSearchExecutors$DirectExecutorService.execute(OpenSearchExecutors.java:341)
	at org.opensearch.transport.InboundHandler.handleException(InboundHandler.java:419)
	at org.opensearch.transport.InboundHandler.handlerResponseError(InboundHandler.java:411)
	at org.opensearch.transport.InboundHandler.messageReceived(InboundHandler.java:159)
	at org.opensearch.transport.InboundHandler.inboundMessage(InboundHandler.java:115)
	at org.opensearch.transport.TcpTransport.inboundMessage(TcpTransport.java:767)
	at org.opensearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:175)
	at org.opensearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:150)
	at org.opensearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:115)
	at org.opensearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:341)
	at org.opensearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:246)
	at org.opensearch.nio.BytesChannelContext.read(BytesChannelContext.java:59)
	at org.opensearch.nio.EventHandler.handleRead(EventHandler.java:152)
	at org.opensearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:167)
	at org.opensearch.nio.NioSelector.handleRead(NioSelector.java:438)
	at org.opensearch.nio.NioSelector.processKey(NioSelector.java:264)
	at org.opensearch.nio.NioSelector.singleLoop(NioSelector.java:191)
	at org.opensearch.nio.NioSelector.runLoop(NioSelector.java:148)
	... 1 more
Caused by: SearchContextMissingException[No search context found for id [1]]
	at org.opensearch.search.SearchService.findReaderContext(SearchService.java:805)
	at org.opensearch.search.SearchService.getShard(SearchService.java:590)
	at org.opensearch.search.SearchService.executeDfsPhase(SearchService.java:505)
	at org.opensearch.action.search.SearchTransportService.lambda$registerRequestHandler$6(SearchTransportService.java:546)
	at org.opensearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:106)
	at org.opensearch.transport.InboundHandler.handleRequest(InboundHandler.java:250)
	at org.opensearch.transport.InboundHandler.messageReceived(InboundHandler.java:133)
	at org.opensearch.transport.InboundHandler.inboundMessage(InboundHandler.java:115)
	at org.opensearch.transport.TcpTransport.inboundMessage(TcpTransport.java:767)
	at org.opensearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:175)
	at org.opensearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:150)
	at org.opensearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:115)
	at org.opensearch.transport.nio.MockNioTransport$MockTcpReadWriteHandler.consumeReads(MockNioTransport.java:341)
	at org.opensearch.nio.SocketChannelContext.handleReadBytes(SocketChannelContext.java:246)
	at org.opensearch.nio.BytesChannelContext.read(BytesChannelContext.java:59)
	at org.opensearch.nio.EventHandler.handleRead(EventHandler.java:152)
	at org.opensearch.transport.nio.TestEventHandler.handleRead(TestEventHandler.java:167)
	at org.opensearch.nio.NioSelector.handleRead(NioSelector.java:438)
	at org.opensearch.nio.NioSelector.processKey(NioSelector.java:264)
	at org.opensearch.nio.NioSelector.singleLoop(NioSelector.java:191)
	at org.opensearch.nio.NioSelector.runLoop(NioSelector.java:148)
	at java.lang.Thread.run(Thread.java:833)
```

**To Reproduce**
Steps to reproduce the behavior:
It will be easier to reproduce this problem if the thread sleeps 1s before setting flag (`deleted.set(true)`). Or you can increase the number of threads and then repeat the test until exception is thrown.

**Expected behavior**
A clear and concise description of what you expected to happen.

**Plugins**
Please list all plugins currently enabled.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Host/Environment (please complete the following information):**
 - OS: [e.g. iOS]
 - Version [e.g. 22]

**Additional context**
Add any other context about the problem here.

--- a/server/src/internalClusterTest/java/org/opensearch/search/pit/DeletePitMultiNodeIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/search/pit/DeletePitMultiNodeIT.java
@@ -8,6 +8,7 @@
 
 package org.opensearch.search.pit;
 
+import org.opensearch.ExceptionsHelper;
 import org.opensearch.action.LatchedActionListener;
 import org.opensearch.action.admin.indices.stats.IndicesStatsRequest;
 import org.opensearch.action.admin.indices.stats.IndicesStatsResponse;
@@ -18,10 +19,14 @@ import org.opensearch.action.search.DeletePitAction;
 import org.opensearch.action.search.DeletePitInfo;
 import org.opensearch.action.search.DeletePitRequest;
 import org.opensearch.action.search.DeletePitResponse;
+import org.opensearch.action.search.SearchPhaseExecutionException;
+import org.opensearch.action.search.SearchResponse;
+import org.opensearch.action.search.ShardSearchFailure;
 import org.opensearch.common.action.ActionFuture;
 import org.opensearch.common.settings.Settings;
 import org.opensearch.common.unit.TimeValue;
 import org.opensearch.core.action.ActionListener;
+import org.opensearch.search.SearchContextMissingException;
 import org.opensearch.search.builder.PointInTimeBuilder;
 import org.opensearch.test.InternalTestCluster;
 import org.opensearch.test.OpenSearchIntegTestCase;
@@ -263,18 +268,23 @@ public class DeletePitMultiNodeIT extends OpenSearchIntegTestCase {
                 try {
                     latch.await();
                     for (int j = 0; j < 30; j++) {
-                        client().prepareSearch()
+                        SearchResponse searchResponse = client().prepareSearch()
                             .setSize(2)
                             .setPointInTime(new PointInTimeBuilder(pitResponse.getId()).setKeepAlive(TimeValue.timeValueDays(1)))
                             .execute()
                             .get();
+                        if (searchResponse.getFailedShards() != 0) {
+                            verifySearchContextMissingException(searchResponse.getShardFailures());
+                        }
                     }
                 } catch (Exception e) {
                     /**
                      * assert for exception once delete pit goes through. throw error in case of any exeption before that.
                      */
                     if (deleted.get() == true) {
-                        if (!e.getMessage().contains("all shards failed")) throw new AssertionError(e);
+                        Throwable t = ExceptionsHelper.unwrapCause(e.getCause());
+                        assertTrue(e.toString(), t instanceof SearchPhaseExecutionException);
+                        verifySearchContextMissingException(((SearchPhaseExecutionException) t).shardFailures());
                         return;
                     }
                     throw new AssertionError(e);
@@ -283,9 +293,9 @@ public class DeletePitMultiNodeIT extends OpenSearchIntegTestCase {
             threads[i].setName("opensearch[node_s_0][search]");
             threads[i].start();
         }
+        deleted.set(true);
         ActionFuture<DeletePitResponse> execute = client().execute(DeletePitAction.INSTANCE, deletePITRequest);
         DeletePitResponse deletePITResponse = execute.get();
-        deleted.set(true);
         for (DeletePitInfo deletePitInfo : deletePITResponse.getDeletePitResults()) {
             assertTrue(pitIds.contains(deletePitInfo.getPitId()));
             assertTrue(deletePitInfo.isSuccessful());
@@ -293,6 +303,13 @@ public class DeletePitMultiNodeIT extends OpenSearchIntegTestCase {
 
         for (Thread thread : threads) {
             thread.join();
+        }
+    }
+
+    private void verifySearchContextMissingException(ShardSearchFailure[] failures) {
+        for (ShardSearchFailure failure : failures) {
+            Throwable cause = ExceptionsHelper.unwrapCause(failure.getCause());
+            assertTrue(failure.toString(), cause instanceof SearchContextMissingException);
         }
     }
 

